File: app/src/main/java/com/kian/pitproject/sudoku/TimeHandler.java
package com.kian.pitproject.sudoku;

import android.os.Handler;
import android.widget.TextView;

public class TimeHandler implements Runnable {

    private TextView timeView;
    private long startTime = 0, millis, savedMillis;
    private Handler timeHandler = new Handler();

    public TimeHandler(TextView timeView) {
        this.timeView = timeView;
        startTime = System.currentTimeMillis();
        timeHandler.postDelayed(this, 0);
        savedMillis = 0;
    }

    @Override
    public void run() {
        millis = savedMillis + System.currentTimeMillis() - startTime;
        int seconds = (int) (millis / 1000);
        int minutes = seconds / 60;
        seconds = seconds % 60;

        timeView.setText(String.format("%d:%02d", minutes, seconds));

        timeHandler.postDelayed(this, 500);
    }

    public long getElapsedTime() {
        return millis;
    }

    public void setSavedMillis(long savedMillis) {
        this.savedMillis = savedMillis;
    }
}

File: app/src/main/java/com/kian/sudoku/TimeHandler.kt
package com.kian.sudoku

import android.os.Handler
import android.widget.TextView

class TimeHandler(private val timeView: TextView) : Runnable {
    private var startTime: Long = 0
    var elapsedTime: Long = 0
        private set
    private var savedMillis: Long
    private val timeHandler = Handler()

    companion object {
        fun formattedTime(t: Long): String {
            val seconds = t % 60
            val minutes = t / 60
            return String.format("%d:%02d", minutes, seconds)
        }
    }

    override fun run() {
        elapsedTime = savedMillis + System.currentTimeMillis() - startTime
        timeView.text = formattedTime(elapsedTime / 1000)
//        var seconds = (elapsedTime / 1000).toInt()
//        val minutes = seconds / 60
//        seconds %= 60
//        timeView.text = String.format("%d:%02d", minutes, seconds)
        timeHandler.postDelayed(this, 500)
    }

    fun setSavedMillis(savedMillis: Long) {
        this.savedMillis = savedMillis
    }

    init {
        startTime = System.currentTimeMillis()
        timeHandler.postDelayed(this, 0)
        savedMillis = 0
    }
}